sys/include/net/gcoap.h

```diff
@@ -378,12 +381,14 @@ typedef struct {
  * @brief  Container for the state of gcoap itself
  */
 typedef struct {
+    mutex_t lock;                       /**< Shares state attributes safely */
     gcoap_listener_t *listeners;        /**< List of registered listeners */
     gcoap_request_memo_t open_reqs[GCOAP_REQ_WAITING_MAX];
                                         /**< Storage for open requests; if first
                                              byte of an entry is zero, the entry
                                              is available */
-    uint16_t last_message_id;           /**< Last message ID used */
+    atomic_uint_least16_t next_message_id;
```

Comment:
@smlng sorry if I was unclear. Yes, I think atomic_uint16_t is the way to go here, and that this can be revisited if we realize that this has become a bottleneck for the system. Using the other types (_least, _fast) may increase the width of the variable on some platforms, which may have consequences in other locations.